pkg/server/io/io.go

```diff
@@ -0,0 +1,317 @@
+		id:         id,
+		dir:        fifos.Dir,
+		stdoutPath: fifos.Out,
+		stderrPath: fifos.Err,
+		stdout:     ioutil.NewWriterGroup(),
+		stderr:     ioutil.NewWriterGroup(),
+	}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+	if c.stdin {
+		c.stdinPath = fifos.In
+	}
+	return c, nil
+}
+
+// Config returns io config.
+func (c *ContainerIO) Config() containerd.IOConfig {
+	return containerd.IOConfig{
+		Terminal: c.tty,
+		Stdin:    c.stdinPath,
+		Stdout:   c.stdoutPath,
+		Stderr:   c.stderrPath,
+	}
+}
+
+// Pipe creates container fifos and pipe container output
+// to output stream.
+func (c *ContainerIO) Pipe() (err error) {
+	var (
+		f           io.ReadWriteCloser
+		set         []io.Closer
+		ctx, cancel = context.WithCancel(context.Background())
+		wg          = &sync.WaitGroup{}
+	)
+	defer func() {
+		if err != nil {
+			for _, f := range set {
+				f.Close()
+			}
+			cancel()
+		}
+	}()
+	if c.stdinPath != "" {
+		// Just create the stdin, only open it when used.
+		if f, err = fifo.OpenFifo(ctx, c.stdinPath, syscall.O_WRONLY|syscall.O_CREAT|syscall.O_NONBLOCK, 0700); err != nil {
+			return err
+		}
+		f.Close()
+	}
+
+	if f, err = fifo.OpenFifo(ctx, c.stdoutPath, syscall.O_RDONLY|syscall.O_CREAT|syscall.O_NONBLOCK, 0700); err != nil {
+		return err
+	}
+	set = append(set, f)
+	wg.Add(1)
+	go func(r io.ReadCloser) {
+		if _, err := io.Copy(c.stdout, r); err != nil {
+			glog.Errorf("Failed to redirect stdout of container %q: %v", c.id, err)
+		}
+		r.Close()
+		c.stdout.Close()
+		wg.Done()
+		glog.V(2).Infof("Finish piping stdout of container %q", c.id)
+	}(f)
+
+	if f, err = fifo.OpenFifo(ctx, c.stderrPath, syscall.O_RDONLY|syscall.O_CREAT|syscall.O_NONBLOCK, 0700); err != nil {
+		return err
+	}
+	set = append(set, f)
+	if !c.tty {
+		wg.Add(1)
+		go func(r io.ReadCloser) {
+			if _, err := io.Copy(c.stderr, r); err != nil {
+				glog.Errorf("Failed to redirect stderr of container %q: %v", c.id, err)
+			}
+			r.Close()
+			c.stderr.Close()
+			wg.Done()
+			glog.V(2).Infof("Finish piping stderr of container %q", c.id)
+		}(f)
+	}
+	c.closer = &wgCloser{
+		wg:     wg,
+		set:    set,
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	return nil
+}
+
+// Attach attaches container stdio.
+func (c *ContainerIO) Attach(stdin io.Reader, stdout, stderr io.WriteCloser) error {
+	if c.closer == nil {
+		return errors.New("container io is not initialized")
+	}
+	var wg sync.WaitGroup
+	key := util.GenerateID()
+	stdinKey := streamKey(c.id, "attach-"+key, Stdin)
+	stdoutKey := streamKey(c.id, "attach-"+key, Stdout)
+	stderrKey := streamKey(c.id, "attach-"+key, Stderr)
+
+	var stdinCloser io.Closer
+	if c.stdinPath != "" && stdin != nil {
+		f, err := fifo.OpenFifo(c.closer.ctx, c.stdinPath, syscall.O_WRONLY|syscall.O_NONBLOCK, 0700)
+		if err != nil {
+			return err
+		}
+		// Also increase wait group here, so that `closer.Wait` will
+		// also wait for this fifo to be closed.
+		c.closer.wg.Add(1)
```

Comment:
Yeah, `ContainerIO.Close` will wait for this. This makes sure that `ContainerIO.Close` won't remove the fifo directory when Attach is still using the stdin pipe.